src/pkg/cached/repository/redis/manager.go

```diff
@@ -0,0 +1,254 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	libcache "github.com/goharbor/harbor/src/lib/cache"
+	"github.com/goharbor/harbor/src/lib/config"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/lib/retry"
+	"github.com/goharbor/harbor/src/pkg/cached"
+	"github.com/goharbor/harbor/src/pkg/repository"
+	"github.com/goharbor/harbor/src/pkg/repository/model"
+)
+
+var _ CachedManager = &manager{}
+
+// CachedManager is the interface combines raw resource manager and cached manager for better extension.
+type CachedManager interface {
+	// Manager is the raw resource manager.
+	repository.Manager
+	// Manager is the common interface for resource cache.
+	cached.Manager
+}
+
+// manager is the cached manager implemented by redis.
+type manager struct {
+	// delegator delegates the raw crud to DAO.
+	delegator repository.Manager
+	// client returns the redis cache client.
+	client func() libcache.Cache
+	// keyBuilder builds cache object key.
+	keyBuilder *cached.ObjectKey
+	// lifetime is the cache life time.
+	lifetime time.Duration
+}
+
+// NewManager returns the redis cache manager.
+func NewManager(m repository.Manager) *manager {
+	return &manager{
+		delegator:  m,
+		client:     func() libcache.Cache { return libcache.Default() },
+		keyBuilder: cached.NewObjectKey(cached.ResourceTypeRepository),
+		lifetime:   time.Duration(config.CacheExpireHours()) * time.Hour,
+	}
+}
+
+func (m *manager) Count(ctx context.Context, query *q.Query) (int64, error) {
+	return m.delegator.Count(ctx, query)
+}
+
+func (m *manager) List(ctx context.Context, query *q.Query) ([]*model.RepoRecord, error) {
+	return m.delegator.List(ctx, query)
+}
+
+func (m *manager) Create(ctx context.Context, repo *model.RepoRecord) (int64, error) {
+	return m.delegator.Create(ctx, repo)
+}
+
+func (m *manager) NonEmptyRepos(ctx context.Context) ([]*model.RepoRecord, error) {
+	return m.delegator.NonEmptyRepos(ctx)
+}
```

Comment:
There are lots of delegation methods in the cacheManager,  if we need to add a method in the repository.Manager, lots of file changes are involved.
Suggest using the embedded interface in manager struct to refactor it.
see https://github.com/goharbor/harbor/blob/09c3d042eac1bd5d2a4c91238f1ab047003d3314/src/pkg/reg/adapter/native/adapter.go#L62

Reply:
Hold on, let's refactor this part if needed in the future.